pkg/controllers/infra/cluster_controller.go

```diff
@@ -369,7 +389,50 @@ func (r *ClusterController) reconcileCNI(scopeCluster *scope.Cluster) error {
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *ClusterController) SetupWithManager(mgr ctrl.Manager) error {
-	return ctrl.NewControllerManagedBy(mgr).
+	c, err := ctrl.NewControllerManagedBy(mgr).
 		For(&clusterv1alpha1.Cluster{}).
-		Complete(r)
+		Build(r)
+	if err != nil {
+		return err
+	}
+
+	if err := c.Watch(
+		&source.Kind{Type: &corev1.Secret{}},
+		handler.EnqueueRequestsFromMapFunc(r.SecretToClusterFunc),
+	); err != nil {
+		return fmt.Errorf("failed adding Watch for Secret to controller manager: %v", err)
+	}
+
+	return nil
+}
+
+func (r *ClusterController) SecretToClusterFunc(o client.Object) []ctrl.Request {
+	s, ok := o.(*corev1.Secret)
+	if !ok {
+		panic(fmt.Sprintf("Expected a Secret but got a %T", o))
+	}
+	var result []ctrl.Request
+
+	// This is not the best way to do this,
+	// but we can not use OwnerReference to link Cluster and Secret,
+	// because the Secret can be referenced by multiple Clusters
+	clusters := &clusterv1alpha1.ClusterList{}
+	if err := r.Client.List(context.TODO(), clusters, client.InNamespace(s.GetNamespace())); err != nil {
+		log.Error(err, "failed to list clusters")
+		return nil
+	}
+
+	sets := sets.NewString()
+	for _, cluster := range clusters.Items {
+		if cluster.Spec.Credential.SecretRef == s.GetName() {
+			nn := client.ObjectKey{Namespace: cluster.Namespace, Name: cluster.Name}
+			if !sets.Has(nn.String()) {
+				result = append(result, ctrl.Request{NamespacedName: nn})
+				sets.Insert(nn.String())
```

Comment:
I am concerned about the performance, can we do as what we have done in other places, add a ownerref or label to secret. Then we do not need to loop over all clusters here

Reply:
I address my current thinking on L416, at least `OwnerReference` is not a good idea.